modules/mono/csharp_script.cpp

```diff
@@ -1696,6 +1696,10 @@ bool CSharpInstance::property_can_revert(const StringName &p_name) const {
 	return (bool)ret;
 }
 
+void CSharpInstance::validate_property(PropertyInfo &p_property) const {
+	// TODO
+}
```

Comment:
Maybe @godotengine/dotnet can help with the implementation.

Reply:
Here's a diff with the implementation (also adds a C# example to the `_validate_property` method documentation):

```diff
diff --git a/doc/classes/Object.xml b/doc/classes/Object.xml
index 90ff9259bc..515a0c98dd 100644
--- a/doc/classes/Object.xml
+++ b/doc/classes/Object.xml
@@ -267,7 +267,8 @@
 			<param index="0" name="property" type="Dictionary" />
 			<description>
 				Override this method to customize existing properties. Every property info goes through this method. The dictionary contents is the same as in [method _get_property_list].
-				[codeblock]
+				[codeblocks]
+				[gdscript]
 				@tool
 				extends Node
 
@@ -280,7 +281,38 @@
 				func _validate_property(property: Dictionary):
 				    if property.name == "number" and not is_number_editable:
 				        property.usage |= PROPERTY_USAGE_READ_ONLY
-				[/codeblock]
+				[/gdscript]
+				[csharp]
+				[Tool]
+				public partial class MyNode : Node
+				{
+				    private bool _isNumberEditable;
+
+				    [Export]
+				    public bool IsNumberEditable
+				    {
+				        get =&gt; _isNumberEditable;
+				        set
+				        {
+				            _isNumberEditable = value;
+				            NotifyPropertyListChanged();
+				        }
+				    }
+
+				    [Export]
+				    public int Number { get; set; }
+
+				    public override void _ValidateProperty(Godot.Collections.Dictionary property)
+				    {
+				        if (property["name"].AsStringName() == PropertyName.Number &amp;&amp; IsNumberEditable)
+				        {
+				            var usage = property["usage"].As&gt;PropertyUsageFlags&lt;() | PropertyUsageFlags.ReadOnly;
+				            property["usage"] = (int)usage;
+				        }
+				    }
+				}
+				[/csharp]
+				[/codeblocks]
 			</description>
 		</method>
 		<method name="add_user_signal">
diff --git a/modules/mono/csharp_script.cpp b/modules/mono/csharp_script.cpp
index 02c5605022..ebc526b864 100644
--- a/modules/mono/csharp_script.cpp
+++ b/modules/mono/csharp_script.cpp
@@ -1658,7 +1658,8 @@ void CSharpInstance::get_property_list(List<PropertyInfo> *p_properties) const {
 		}
 	}
 
-	for (const PropertyInfo &prop : props) {
+	for (PropertyInfo &prop : props) {
+		validate_property(prop);
 		p_properties->push_back(prop);
 	}
 }
@@ -1697,7 +1698,21 @@ bool CSharpInstance::property_can_revert(const StringName &p_name) const {
 }
 
 void CSharpInstance::validate_property(PropertyInfo &p_property) const {
-	// TODO
+	ERR_FAIL_COND(!script.is_valid());
+
+	Variant property_arg = (Dictionary)p_property;
+	const Variant *args[1] = { &property_arg };
+
+	Variant ret;
+	Callable::CallError call_error;
+	GDMonoCache::managed_callbacks.CSharpInstanceBridge_Call(
+			gchandle.get_intptr(), &SNAME("_validate_property"), args, 1, &call_error, &ret);
+
+	if (call_error.error != Callable::CallError::CALL_OK) {
+		return;
+	}
+
+	p_property = PropertyInfo::from_dict(property_arg);
 }
 
 bool CSharpInstance::property_get_revert(const StringName &p_name, Variant &r_ret) const {

```